### PR TITLE
automatically assign cgroupDriver/resolverConfig in Kubelet config file

### LINKED ISSUE #874: remove the docker cgroup driver detection
#### old issue was tracking cgroup driver detection for docker and other CRs

this is targeting 1.22.

remove the docker cgroupdriver detection in kubeadm:
- [x] k/k PR:
https://github.com/kubernetes/kubernetes/pull/99647

- [x] k/website PR:
sent website PR to update these pages after the change:
https://github.com/kubernetes/website/blob/281df0c16b0ce403bfa1442388794c13fbcc9d27/content/en/docs/setup/production-environment/tools/kubeadm/install-kubeadm.md
https://github.com/kubernetes/website/blob/253d6ab66db2972a0af21c27af77458fe3a841f9/content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md
https://github.com/kubernetes/website/blob/cfe5edfd3112bc4e94160dc09fedd4900ea14608/content/en/docs/setup/production-environment/container-runtimes.md
https://github.com/kubernetes/website/blob/401c4fc0fae978d73894418a710b14c5a0474098/content/en/docs/tasks/administer-cluster/reserve-compute-resources.md

- [x] pin the cgroup driver for docker / containerd to "systemd" in kinder. should be done [here](
https://github.com/kubernetes/kubeadm/blob/daf8cabc4ed98f90a1adc5a7251a629e9874eb03/kinder/pkg/build/alter/alter.go#L280); information about setting the driver for CRs is [here](https://kubernetes.io/docs/setup/production-environment/container-runtimes/)
https://github.com/kubernetes/kubeadm/pull/2495

--------------------
old description:

this was already done for docker:
https://github.com/kubernetes/kubeadm/issues/844

> When kubernetes/kubernetes#63887 merges, we should set the kubelet's --cgroup-driver flag to either systemd or cgroupfs conditionally at kubeadm init time, in cmd/kubeadm/app/phases/kubelet/flags.go.
This instead of doing it statically in the deb/rpm packages.
xref related issue: #822

we need to detect the driver for other CRI as well.

Docker PR:
kubernetes/kubernetes#64347

Document outlining some of the previous proposals and issues for handling this in the kubelet:
https://github.com/filbranden/community/blob/009f067c6bc106d062a416d0a1537de5ffa8c023/contributors/design-proposals/node/cgroup-driver-kubelet-and-runtime.md

### PR DESCRIPTION
**Pull request by chenzhiwei**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

`--cgroup-driver` command line argument was deprecated and moved to kubelet config file, this PR is used to automatically assign cgroupDriver in Kubelet config file.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubeadm/issues/874

**Special notes for your reviewer**:

/area kubeadm
sig/cluster-lifecycle

/assign @yagonobre
/assign @neolit123 

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: apply the detected cgroup-driver and systemd-resolved as values in the KubeletConfiguration instead of flags to the kubelet binary.
```

### CODE REVIEW

File: cmd/kubeadm/app/componentconfigs/defaults.go
Comment (neolit123) [inline cmd/kubeadm/app/componentconfigs/defaults.go]: please change this line to:
```
return errors.Wrap(err, "cannot automatically assign a cgroupDriver value in the KubeletConfiguration file")
```

**Review by neolit123 — COMMENTED** [review]
@kubernetes/sig-cluster-lifecycle-pr-reviews 

@chenzhiwei thank you for this PR. added one minor comment.
this change is correct - we need to use the config instead of flags.

but if we are moving away from the flag this has to be removed:
https://github.com/kubernetes/kubernetes/blob/e47fc04adc39c06bcdd6d7dc74ee881d8740a81a/cmd/kubeadm/app/phases/kubelet/flags.go#L79-L83

as part of the same PR.

File: cmd/kubeadm/app/componentconfigs/defaults.go
Comment (neolit123) [reply cmd/kubeadm/app/componentconfigs/defaults.go]: EDIT actually this function does not return an error...
please use:
```
klog.Warningf("cannot automatically assign a cgroupDriver value in the KubeletConfiguration file: %v", err)
```

**Comment by chenzhiwei** [comment]
@neolit123 Thanks, I rebased the code and addressed your comments.

File: cmd/kubeadm/app/componentconfigs/defaults.go
Comment (rosti) [inline cmd/kubeadm/app/componentconfigs/defaults.go]: This code needs to be executed only if the container runtime is Docker. It would break containerd, CRI-O, etc.
To do that we need to get the `NodeRegistrationOptions` propagated to this func as an additional parameter as well.

**Review by rosti — COMMENTED** [review]
Thanks for tackling this @chenzhiwei !

File: cmd/kubeadm/app/componentconfigs/defaults.go
Comment (neolit123) [reply cmd/kubeadm/app/componentconfigs/defaults.go]: ^ yes, this is true.

**Comment by chenzhiwei** [comment]
@neolit123 @rosti I updated the PR and added a condition check.

**Review by dixudx — APPROVED** [review]
/lgtm

File: cmd/kubeadm/app/componentconfigs/defaults.go
Comment (detiber) [inline cmd/kubeadm/app/componentconfigs/defaults.go]: stdlib imports should be grouped separately and before other imports.

Comment (chenzhiwei) [reply cmd/kubeadm/app/componentconfigs/defaults.go]: Thank you for reviewing this PR, I just fixed this. (my editor automatically moved this to the end)

**Review by timothysc — APPROVED** [review]
/approve 

let others finish their review for lgtm

**Review by neolit123 — COMMENTED** [review]
/lgtm
/hold
/test pull-kubernetes-e2e-kind

**Comment by neolit123** [comment]
/lgtm cancel
@chenzhiwei 

we need to wait for your other PR to merge:
https://github.com/kubernetes/kubernetes/pull/77032

and re-run `/test pull-kubernetes-e2e-kind`.

right now this change fails on a real cluster:
> Pulling: W0427 11:56:28.574843     104 defaults.go:82] cannot automatically assign a cgroupDriver value in the KubeletConfiguration file: cannot execute 'docker info': exit status 1
time="11:56:28" level=info msg="Pulling image: W0427 11:56:28.574843     104 defaults.go:82] cannot automatically assign a cgroupDriver value in the KubeletConfiguration file: cannot execute 'docker info': exit status 1 ..."

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/77059/pull-kubernetes-e2e-kind/1122105836089380864/build-log.txt

**Comment by chenzhiwei** [comment]
I built this PR code and tested locally, I ran `kubeadm init phase kubelet-start` and it generated corrected `kubeadm-flags.env` and `config.yaml`.

The PR #77032 is using a new method to detect cgroup driver, the old method should still work.

@neolit123 Could you explain more about these two PRs? I still don't know why they have relationship.

**Comment by neolit123** [comment]
> Could you explain more about these two PRs? I still don't know why they have relationship.

we they are:
- both by you :)
- modify the kubeadm cgroup driver logic

> The PR #77032 is using a new method to detect cgroup driver, the old method should still work.

this is true.

but this test needs to pass on this PR before we can merge it:
/test pull-kubernetes-e2e-kind

**Comment by chenzhiwei** [comment]
When running the test, I also encounter following errors, need more time to investigate.

(`kubeadm init` works well in my vagrant VM: Ubuntu 18.04)

```
make test GOFLAGS=-v WHAT=./cmd/kubeadm/app/util/config/
...
...
...
    --- FAIL: TestMigrateOldConfigFromFile/InitConfiguration_only_gets_migrated (2.55s)
        common_test.go:389: length mismatch between resulting gvks and expected kinds:
            	len(gvks)=3
            	len(expectedKinds)=2
    --- FAIL: TestMigrateOldConfigFromFile/ClusterConfiguration_only_gets_migrated (2.81s)
        common_test.go:389: length mismatch between resulting gvks and expected kinds:
            	len(gvks)=3
            	len(expectedKinds)=2
    --- PASS: TestMigrateOldConfigFromFile/JoinConfiguration_only_gets_migrated (0.00s)
    --- FAIL: TestMigrateOldConfigFromFile/Init_+_Cluster_Configurations_are_migrated (2.96s)
        common_test.go:389: length mismatch between resulting gvks and expected kinds:
            	len(gvks)=3
            	len(expectedKinds)=2
    --- FAIL: TestMigrateOldConfigFromFile/Init_+_Join_Configurations_are_migrated (3.38s)
        common_test.go:389: length mismatch between resulting gvks and expected kinds:
            	len(gvks)=4
            	len(expectedKinds)=3
    --- FAIL: TestMigrateOldConfigFromFile/Cluster_+_Join_Configurations_are_migrated (2.46s)
        common_test.go:389: length mismatch between resulting gvks and expected kinds:
            	len(gvks)=4
            	len(expectedKinds)=3
    --- FAIL: TestMigrateOldConfigFromFile/Init_+_Cluster_+_Join_Configurations_are_migrated (2.97s)
        common_test.go:389: length mismatch between resulting gvks and expected kinds:
            	len(gvks)=4
            	len(expectedKinds)=3
    --- FAIL: TestMigrateOldConfigFromFile/component_configs_are_not_migrated (2.67s)
        common_test.go:389: length mismatch between resulting gvks and expected kinds:
            	len(gvks)=4
            	len(expectedKinds)=3
```

File: cmd/kubeadm/app/componentconfigs/defaults.go
Comment (neolit123) [inline cmd/kubeadm/app/componentconfigs/defaults.go]: this will not work in all setups, but also dynamically detecting the cgroup driver in DefaultKubeletConfiguration is not the right place for this.

we use the tool `kind` for e2e testing. kind uses `kubeadm config images list`:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/77059/pull-kubernetes-e2e-kind/1122122067366383616/build-log.txt
`kubeadm config images list` ends up calling `GetCgroupDriverDocker` but Docker is not running in the build container. we should avoid detecting a cgroup driver as part of the defaulting functions.

unfortunately the required change is much more complicated.
for kubeadm init the driver needs to be detected before the KubeletConfiguration file is written to disk:
https://github.com/kubernetes/kubernetes/blob/a404238d1c90ee00827624249b9223e7644e3f99/cmd/kubeadm/app/cmd/phases/init/kubelet.go#L71

for kubeadm join the KubeletConfiguration is downloaded from the cluster and written to disk:
https://github.com/kubernetes/kubernetes/blob/a404238d1c90ee00827624249b9223e7644e3f99/cmd/kubeadm/app/cmd/phases/join/kubelet.go#L134-L137
the KubelConfiguration that was downloaded needs to be modified with the detected cgroup driver before written to disk.

i'm starting to think that we should not do this refactor as is, but instead think how to translate all the flags from:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubelet/flags.go
to the KubeletConfiguration and not only `cgroupDriver`.

/hold

**Comment by neolit123** [comment]
> When running the test, I also encounter following errors

this works:
```
~/go/src/k8s.io/kubernetes$ go test -v ./cmd/kubeadm/app/util/config/...
```

File: cmd/kubeadm/app/componentconfigs/defaults.go
Comment (chenzhiwei) [reply cmd/kubeadm/app/componentconfigs/defaults.go]: Yes, the kubelet config content is generated too early, it is better to generate before writing it to disk, just like the `kubeadm-flags.env` file.

**Comment by chenzhiwei** [comment]
@neolit123 @timothysc @detiber 

I updated this PR to update the cgroupDriver when before writing config file to disk, could you review the change again?

**Review by neolit123 — COMMENTED** [review]
/approve cancel
will have a better look later today.
thanks for the update.

File: cmd/kubeadm/app/phases/kubelet/config.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/config.go]: kubeadm join does not use this function when writing the file to disk.
it uses:
https://github.com/kubernetes/kubernetes/blob/a404238d1c90ee00827624249b9223e7644e3f99/cmd/kubeadm/app/cmd/phases/join/kubelet.go#L134-L137

then it uses this private function:
`writeConfigBytesToDisk`
https://github.com/kubernetes/kubernetes/blob/9c75270b2be4dcc39e31edb34b76e8f604aaf91a/cmd/kubeadm/app/phases/kubelet/config.go#L157

so the driver detection on join will still result in a flag instead of a config value.

both init and join end up calling `writeConfigBytesToDisk`, but i don't think we should have the driver detection in there.

this needs a much larger refactor.
instead of moving only the cgroup driver detection away from `cmd/kubeadm/app/phases/kubelet/flags.go` my idea was to move all of the flags that can be under the config from here:
https://github.com/kubernetes/kubernetes/blob/e47fc04adc39c06bcdd6d7dc74ee881d8740a81a/cmd/kubeadm/app/phases/kubelet/flags.go#L73-L115

into a new intermediate step that applies to the KubeletConfiguration that is written to disk on both init and join.

Comment (rosti) [reply cmd/kubeadm/app/phases/kubelet/config.go]: @neolit123 you are right. Currently there are a couple of flags in that code `cgroup-driver` and `resolv-conf` with the latter one being clearly host specific. We need to remove them from there and allow for host specific fields of the kubelet component config to be patched locally (either on create or on download of the config).

**Comment by chenzhiwei** [comment]
Will start the new refactor and push new change.

**Comment by chenzhiwei** [comment]
@neolit123 @timothysc @rosti The refactor is a bit complex now, what I think is:

1. Introduce a new function `buildDynamicKubeletConfig`

    `WriteConfigToDisk` and `DownloadConfig` call this function to dynamically generate kubelet config.yaml file.

    This can cover various cases:

    * init: `kubeadm init`
    * join: `kubeadm join`
    * upgrade: `kubeadm upgrade`
    * download: `kubeadm alpha kubelet config download`

    For last three cases, they all call the DownloadConfig.

2. init & join are easy to implement

3. for download, introduce a new parameter `--cri-socket`

    When users run download, kubeadm gets CRISocket from command line, if not in command line, then automatically detect it.

    code logic in: cmd/kubeadm/app/cmd/alpha/kubelet.go

4. [WIP] for upgrade, get the old kubelet config.yaml and merge with the new downloaded one

    This part needs more refactor about the DownloadConfig, need to introduce two new functions: DownloadAndMergeConfig, mergeKubeletConfig and getCurrentKubeletConfig.

    DownloadConfig contains download, dynamically generate and save to disk.

    DownloadAndMergeConfig contains download, merge and save to disk.

**Comment by chenzhiwei** [comment]
@neolit123 @timothysc @rosti could you review this PR again? I updated with more changes.

File: cmd/kubeadm/app/componentconfigs/config.go
Comment (neolit123) [inline cmd/kubeadm/app/componentconfigs/config.go]: do we need this change in the master branch?

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/alpha/kubelet.go]: this mutates the downloaded KubeletConfiguration so that the user can override CRI socket or use autodetection.
this might be the right thing to do.

@rosti PTAL.

File: cmd/kubeadm/app/componentconfigs/config.go
Comment (neolit123) [inline cmd/kubeadm/app/componentconfigs/config.go]: i think we can avoid the addition of this function - i.e. unroll it in the single caller.

File: cmd/kubeadm/app/phases/kubelet/config.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/config.go]: the function should be a no-op for Windows.

i suggest you create:
```
cmd/kubeadm/app/phases/kubelet/config.go
cmd/kubeadm/app/phases/kubelet/config_windows.go
cmd/kubeadm/app/phases/kubelet/config_unix.go
```

and have this function be different between the OS-es.

cc @ksubrmnn

File: cmd/kubeadm/app/phases/kubelet/config.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/config.go]: the name conflates with the feature called "dynamic kublet configuration" (see dynamic.go)
i suggest a rename: `buildDynamicDefaults()`

**Review by neolit123 — COMMENTED** [review]
@chenzhiwei thanks for the update.
added further comments.

**Review by chenzhiwei — COMMENTED** [review]
made a new change to fix the comment issues.

File: cmd/kubeadm/app/phases/kubelet/config.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/config.go]: please change the error message just to:
`received an invalid KubeletConfiguration`

File: cmd/kubeadm/app/phases/kubelet/config_unix.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/config_unix.go]: 2019
because it's a new file.
[1]

File: cmd/kubeadm/app/phases/kubelet/config_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/config_windows.go]: [1]

File: cmd/kubeadm/app/phases/kubelet/config_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/config_windows.go]: please add note under `// buildDynamicDefaults returns ...`
`// This function is a no-op on Windows`

**Review by neolit123 — COMMENTED** [review]
thanks for the update @chenzhiwei 
added some more minor comments, but this is almost ready.

**Comment by neolit123** [comment]
@chenzhiwei 
^ this LGTM once the above test passes.

please add a release note under "Does this PR introduce a user-facing change?:"
instead of `NONE`

> kubeadm: apply the detected `cgroup-driver` and `systemd-resolved` as values in the KubeletConfiguration instead of flags to the kubelet binary.

**Comment by chenzhiwei** [comment]
@neolit123 Rebased the code, hope this time `pull-kubernetes-e2e-kind` can use the right commit to test.

**Comment by rosti** [comment]
/assign
to TAL later today.

**Comment by neolit123** [comment]
compile is failing:
> ERROR: /home/prow/go/src/github.com/kubernetes/kubernetes/cmd/kubeadm/app/cmd/phases/upgrade/node/BUILD:3:1: GoCompile cmd/kubeadm/app/cmd/phases/upgrade/node/linux_amd64_race_stripped/go_default_library%/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/upgrade/node.a failed (Exit 1): compile failed: error executing command

**Comment by chenzhiwei** [comment]
@neolit123 The code base was changed a lot, I need to read the new code and do a full rebase. Thanks.

**Comment by chenzhiwei** [comment]
@neolit123 @rosti @timothysc Hi, do I need any additional update on this PR?

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (kad) [inline cmd/kubeadm/app/cmd/alpha/kubelet.go]: it can be simplified to `return kubeadmruntime.DetectCRISocket()`
In overall, I don't think we need that function if it is called only in one place. It is easier to inline this if/else in the `newCmdKubeletConfigDownload`

File: cmd/kubeadm/app/phases/kubelet/config_unix.go
Comment (kad) [inline cmd/kubeadm/app/phases/kubelet/config_unix.go]: tests?

**Review by kad — COMMENTED** [review]
It feels like we are changing quite a lot in functionality, but also reducing amount of unit tests, which doesn't feel right.

/cc @bart0sh please check this PR.

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (bart0sh) [inline cmd/kubeadm/app/cmd/alpha/kubelet.go]: criSocketStr is always empty string here, right? Why not to pass it explicitly, i.e. ```criSocket, err := getCRISocket("")``` ?

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (bart0sh) [inline cmd/kubeadm/app/cmd/alpha/kubelet.go]: This condition will always be false as this function is always called with an empty string. I agree with @kad here - we don't need this function.

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (chenzhiwei) [reply cmd/kubeadm/app/cmd/alpha/kubelet.go]: if you read the full code you will know how does kubelet assign value to criSocketStr.

Please read line 129:

> cmdutil.AddCRISocketFlag(cmd.Flags(), &criSocketStr)

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (chenzhiwei) [reply cmd/kubeadm/app/cmd/alpha/kubelet.go]: Please read line 129:

> Please read line 129:

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (chenzhiwei) [reply cmd/kubeadm/app/cmd/alpha/kubelet.go]: This is used to allow customize the CRI Socket when downloading kubelet config file.

The getCRISocket function is very like the getKubeletVersion function, it is used to make code logic simple and easy to read.

File: cmd/kubeadm/app/phases/kubelet/config_unix.go
Comment (chenzhiwei) [reply cmd/kubeadm/app/phases/kubelet/config_unix.go]: will add it.

**Review by rosti — COMMENTED** [review]
Thanks for picking this up @chenzhiwei !
I think, that this is going in a different path, than what I envisioned. The kubelet component config should be handled by the `componentconfig` module. This means, that patching the component config for the local node (via the NodeRegistrationOptions & local checks) should be done immediately after download of the config, but before it being made available through the internal `ClusterConfiguration`.
Then we should simply serialize that into the kubelet config file.
In that respect `DownloadConfig` should be deleted.

**Comment by chenzhiwei** [comment]
in this approach, componentconfig handles the initial config, and later we
have build dynamic config and the write to disk.

there are various places that construct and write file to disk, can
componentconfig handle all these cases?

**Comment by rosti** [comment]
> in this approach, componentconfig handles the initial config, and later we have build dynamic config and the write to disk. there are various places that construct and write file to disk, can componentconfig handle all these cases?

You should always be able to get the active kubelet component config, provided that you have an internal `InitConfiguration` or `ClusterConfiguration`.

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (kad) [reply cmd/kubeadm/app/cmd/alpha/kubelet.go]: in any way, construct 

```go
	criSocket, err := kubeadmruntime.DetectCRISocket()
	if err != nil {
		return "", err
	}
	return criSocket, nil
```

can be reduced to 
```go
	return kubeadmruntime.DetectCRISocket()
```

**Review by chenzhiwei — COMMENTED** [review]
there are too many changes and too many review comments.

I need to spend much time to do a rebase.

Can we quickly review this PR?

I can create test case for the private function after this PR is merged.

File: cmd/kubeadm/app/phases/kubelet/config_unix.go
Comment (chenzhiwei) [reply cmd/kubeadm/app/phases/kubelet/config_unix.go]: @kad could you give some tips about writing test case for this private function? I tried, but still not figure out yet.

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (rosti) [reply cmd/kubeadm/app/cmd/alpha/kubelet.go]: Actually, we don't need the extra var `criSocket` here. We can reuse `criSocketStr` and even have it renamed to `criSocket` only.

File: cmd/kubeadm/app/phases/kubelet/config_unix.go
Comment (rosti) [reply cmd/kubeadm/app/phases/kubelet/config_unix.go]: @chenzhiwei please see the original tests of `buildKubeletArgMap`, but basically you need to supply some mocks in here (for `utilsexec` and `PidOf`) and then test with these.

**Review by rosti — COMMENTED** [review]
@chenzhiwei do you have the time to continue the work here?
If you have any questions, don't hesitate to ping me.

**Comment by chenzhiwei** [comment]
@rosti Thank you, I will submit another change as soon as possible.

**Comment by chenzhiwei** [comment]
@rosti sorry, I can't continue on this PR now. Will close this PR.

**Pull request closed without merging**
